contracts/nois-payment/src/contract.rs

```diff
@@ -0,0 +1,260 @@
+    ensure_eq, entry_point, to_binary, BankMsg, Coin, CosmosMsg, Deps, DepsMut, Env, MessageInfo,
+    QueryResponse, Response, StdResult, Uint128, WasmMsg,
+};
+
+use crate::error::ContractError;
+use crate::msg::{ConfigResponse, ExecuteMsg, InstantiateMsg, NoisSinkExecuteMsg, QueryMsg};
+use crate::state::{Config, CONFIG};
+
+/// Constant defining the denom of the Coin to be used for payment
+const PAYMENT_DENOM: &str = "unois";
+
+#[entry_point]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    let nois_sink_addr = deps
+        .api
+        .addr_validate(&msg.nois_sink)
+        .map_err(|_| ContractError::InvalidAddress)?;
+    let nois_com_pool_addr = deps
+        .api
+        .addr_validate(&msg.nois_com_pool_addr)
+        .map_err(|_| ContractError::InvalidAddress)?;
+    CONFIG.save(
+        deps.storage,
+        &Config {
+            community_pool: nois_com_pool_addr,
+            sink: nois_sink_addr,
+            gateway: info.sender.to_owned(),
+        },
+    )?;
+    Ok(Response::new()
+        .add_attribute("action", "instantiate")
+        .add_attribute("nois_sink", msg.nois_sink)
+        .add_attribute("nois_community_pool", msg.nois_com_pool_addr)
+        .add_attribute("nois_gateway", info.sender))
+}
+
+#[entry_point]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    match msg {
+        ExecuteMsg::Pay {
+            burn,
+            community_pool,
+            relayer,
+        } => execute_pay(deps, info, env, burn, community_pool, relayer),
+    }
+}
+
+#[entry_point]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<QueryResponse> {
+    let response = match msg {
+        QueryMsg::Config {} => to_binary(&query_config(deps)?)?,
+    };
+    Ok(response)
+}
+
+fn query_config(deps: Deps) -> StdResult<ConfigResponse> {
+    let config = CONFIG.load(deps.storage)?;
+    Ok(config)
+}
+
+fn execute_pay(
+    deps: DepsMut,
+    info: MessageInfo,
+    _env: Env,
+    burn: Uint128,
+    community_pool: Uint128,
+    relayer: (String, Uint128),
+) -> Result<Response, ContractError> {
+    let funds = info.funds;
+
+    // Make sure the caller is gateway to make sure malicious people can't drain someone else's payment balance
+    let gateway = CONFIG.load(deps.storage).unwrap().gateway;
+    ensure_eq!(info.sender, gateway, ContractError::Unauthorized);
+
+    // Check there are no funds. Not a payable Msg
+    if !funds.is_empty() {
+        return Err(ContractError::DontSendFunds);
+    }
+    // Check relayer addr is valid
+    deps.api
+        .addr_validate(relayer.0.as_str())
+        .map_err(|_| ContractError::InvalidAddress)?;
+
+    // Burn
+    let mut out_msgs: Vec<CosmosMsg> = vec![WasmMsg::Execute {
+        contract_addr: CONFIG.load(deps.storage).unwrap().sink.to_string(),
```

Comment:
We should not load the config multiple times in one function. You can just get the full `let config = ...` in the top and then use the values in multiple places.